### ISSUE TITLE
Assign report to me should check for existing assignee

### ISSUE THREAD

**Issue by J-tt** [mastodon/mastodon#20570]
### Pitch

The assign button should make a request in the background to check if the assigned has changed since the page load, and if so, warn the user.

### Motivation

When moderators are quickly processing reports, it’s common for multiple mods to try to self assign the same report which leads to multiple actions being taken on one report.

This change would hugely improve the report assign flow and makes it workable for larger servers, with relatively little required changes.

**Comment by goldjg** [comment]
+1 👍🏻

**Comment by Gargron** [comment]
In a future update i would like to add automatic round robin assignment, that could improve the workflow.

**Comment by goldjg** [comment]
So it would automatically choose a moderator and auto assign when a report is raised?

Regards,
Graham

**Comment by ThisIsMissEm** [comment]
> So it would automatically choose a moderator and auto assign when a report is raised? 

Yes, that'd be the idea. The assignments feature was originally intended for long-running reports, where it wasn't a clean cut yes/no handling of the report, so if you needed to chase someone for additional info, you'd assign it to yourself to ensure that someone else doesn't jump on it.

We could likely use some sort of "if not modified since" on all report actions, such that one moderator doesn't clobber another moderator's actions.

**Comment by ThisIsMissEm** [comment]
In order to implement this, I think we'd need a sort of "on call schedule" for moderators? I'm not sure if that's something really for core, or if it's best left to external tools like, say, pager duty?

**Comment by goldjg** [comment]
In this case, all it needs to do is check the assignee field of the report, if it’s null/empty, assign it to the logged in admin who is clicking that button otherwise if it has a value just display a dialog saying it is already assigned to another admin. Ideally it would tell who that admin is and give you a sort of “are you sure?” choice - either continue to assign to me or cancel and leave it with the existing assignee.

**Comment by quintessence** [comment]
Opinion: I think this could be complex enough that a 3rd party integration with an incident management tool would actually be a great benefit here rather than building it into Mastodon. Building a tool like this not only requires an understanding of who the admin users are, which Mastodon and its peers already has, but also requires a complex understanding of calendars and scheduling as well as multiple notification channels (like email or Discord / Slack) depending on a workflow. And then, of course, maintaining those additions to the codebase.

In my opinion it'd make sense to make an integration with a common incident management tool (or two, if there's bandwidth)  so that the overhead is on a company / OSS project that focuses on incidents.
